### PR TITLE
feature: Allow creation for `"any"` and `"untagged"` as EVCs

### LINKED ISSUE #219: Allow creation of EVCs for untagged UNI - support matching packets without a VLAN tag
Hi,

We should enhance mef_eline to allow the creation of EVCs for untagged UNIs, i.e., matching packets without a VLAN tag.

According to OpenFlow spec 1.3:

> Omitting the OFPXMT_OFB_VLAN_VID field specifies that a flow entry should match packets regardless of whether they contain the corresponding tag. Special values are defined below for the VLAN tag to allow matching of packets with any tag, independent of the tag’s value, and to supports matching packets without a VLAN tag. The special values defined for OFPXMT_OFB_VLAN_VID are:
> /* The VLAN id is 12-bits, so we can use the entire 16 bits to indicate
>  * special conditions.
>  */
> enum ofp_vlan_id {
>     OFPVID_PRESENT = 0x1000, /* Bit that indicate that a VLAN id is set */
>     OFPVID_NONE    = 0x0000, /* No VLAN id was set. */
> };

in our case, creating an EPL will work for untagged traffic, but it will end up matching much more than just untagged traffic: it will match everything in the interface.

If we try to create the EVC, I couldn't find an option to use untagged:

```
root@5cf51067a6e2:/# curl -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc -d '{"name": "vlan100", "dynamic_backup_path": true, "uni_a": {"tag": {"value": 0, "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:01:1"}, "uni_z": {"tag": {"value": 0, "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:03:1"}}'
{"code":400,"description":"VLAN tag 0 is not available in uni_a","name":"Bad Request"}
root@5cf51067a6e2:/# curl -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc -d '{"name": "vlan100", "dynamic_backup_path": true, "uni_a": {"tag": null, "interface_id": "00:00:00:00:00:00:00:01:1"}, "uni_z": {"tag": null, "interface_id": "00:00:00:00:00:00:00:03:1"}}'
{"code":400,"description":"The request body contains invalid API data. None for not nullable for field uni_a/tag.","name":"Bad Request"}
root@5cf51067a6e2:/# curl -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc -d '{"name": "vlan100", "dynamic_backup_path": true, "uni_a": {"tag": {}, "interface_id": "00:00:00:00:00:00:00:01:1"}, "uni_z": {"tag": {}, "interface_id": "00:00:00:00:00:00:00:03:1"}}'
{"code":400,"description":"The request body contains invalid API data. 'tag_type' is a required property for field uni_a/tag.","name":"Bad Request"}
```

### PR DESCRIPTION
**Pull request by Alopalao**
Closes #219

### Summary
Removed `required` property from `#/components/schemas/Tag`
Allow `string` values in `uni.tag.value`, (only "any" and "untagged" for now)

### Local Tests
New `uni` allowed in EVCs requests:
```
{"uni_a": {"tag": {"value": "any", "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:01:1"}}

{"uni_z": {"tag": {"value": "untagged", "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:02:2"}}'
```
Comparing results to OVS
Case: Matching only packets with a VLAN tag regardless of its value (`any` to `"4096/4096"`)
OVS flow created:
`cookie=0xaae4abdd793f8e41, duration=15.738s, table=0, n_packets=0, n_bytes=0, send_fTlow_rem priority=20000,in_port="s1-eth1",vlan_tci=0x1000/0x1000 actions=push_vlan:0x88a8,output:"s1-eth3"`

Case: Matching only packets without a VLAN tag (`untagged` to `0`)
OVS flow created:
`cookie=0xaaa16d05a1189a4f, duration=2.930s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=20000,in_port="s1-eth1",vlan_tci=0x0000/0x1fff actions=push_vlan:0x88a8,output:"s1-eth3"`

Packets reception tested with mininet `ping` feature.
Cases are working with `CONSISTENCY_CHECK` enabled.

### End-to-End Tests
These test include the e2e test recently [committed](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/210).
Also this new [PR](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/213).
```
tests/test_e2e_01_kytos_startup.py ..                                    [  0%]
tests/test_e2e_05_topology.py ..................                         [  9%]
tests/test_e2e_10_mef_eline.py ..........ss.....x.....x...............   [ 27%]
tests/test_e2e_11_mef_eline.py .....                                     [ 29%]
tests/test_e2e_12_mef_eline.py .....xx.                                  [ 33%]
tests/test_e2e_13_mef_eline.py .....xxXx......xxXx.XXxX.xxxx..x......... [ 52%]
...                                                                      [ 53%]
tests/test_e2e_14_mef_eline.py x                                         [ 53%]
tests/test_e2e_15_mef_eline.py ..                                        [ 54%]
tests/test_e2e_20_flow_manager.py ....................                   [ 64%]
tests/test_e2e_21_flow_manager.py ..                                     [ 64%]
tests/test_e2e_22_flow_manager.py ...............                        [ 71%]
tests/test_e2e_23_flow_manager.py ..............                         [ 78%]
tests/test_e2e_30_of_lldp.py ....                                        [ 80%]
tests/test_e2e_31_of_lldp.py ...                                         [ 81%]
tests/test_e2e_32_of_lldp.py ...                                         [ 82%]
tests/test_e2e_40_sdntrace.py .....                                      [ 85%]
tests/test_e2e_41_kytos_auth.py ........                                 [ 88%]
tests/test_e2e_50_maintenance.py ........................                [100%]
```

### CODE REVIEW

**Review by viniarck — COMMENTED** [review]
@Alopalao thanks for this PR. 

On issue #219, we'll need to align the approach that we're going go for on `of_core` as well just so this change works as intended. We'll need to give a semantics of probably value 0 or None to give the semantics of any VLAN tag since currently, `of_core` [is using an logical OR](https://github.com/kytos-ng/of_core/blob/master/v0x04/match_fields.py#L45) with `VlanId.OFPVID_PRESENT`

@italovalcy did an initial assessment about this, I'll suggest that you discuss with him and document on that issue the approach we'll go for, and then after then you can adapt this PR here and map the rest of the changes. 

Thank you

**Comment by Alopalao** [comment]
As discussed:
For `tag.value` to be `0` and `-1` are not the best option. They are replaced with `any` and `untagged` respectively.
On commit [d6772b39f12f35aa3493b498da74494b6aaa9801](https://github.com/kytos-ng/mef_eline/pull/258/commits/d6772b39f12f35aa3493b498da74494b6aaa9801):
- `mef_eline` is change so it supports `strings` in `uni.tag.value`. For now, the string values can only be `any` and `untagged`. This restriction is validated on `TAGDoc`.
- For `any` the action `{"action_type": "set_vlan", "vlan_id": "any"}` is not being added to `flow_mod`s. `nni` and `direct_uni` missing changes (testing).

**Comment by viniarck** [comment]
@Alopalao, thanks for anticipating some PRs, as we've just briefly talked on the planning meeting today, let's also make sure that the approach we end up is aligned with MEF vlan range, here's some initial points to include in the discussion:

1) In the short-term, `dl_vlan` with a `mask` will need to be supported on https://github.com/kytos-ng/mef_eline/issues/18, currently `of_core` is structured to handle either an `int` or a string containing an integer and its mask [`"int/int"`](https://github.com/kytos-ng/of_core/blob/master/v0x04/match_fields.py#L43). 0's in the mask means don't care, so `"an int/0"` means any vlan. `flow_manager` exposes a relatively low-level OpenFlow API, so we should model in a way that facilitates for building the messages. On higher level APIs, such as `mef_eline`'s since it's also driving the range of VLANs, we could use higher level values there such as `"untagged"` or `"any"` but it'll need to translate to what flow_manager expects, and it also needs to be able to express a range in the vlan range feature, let's assess the trade offs and try to keep as simple and compatible as we can.

2) Regarding `"untagged"`, I've noticed that OvS - the default switches platform that we use on mininet [`uses vlan 0, vlan_tci=0 Match only packets without an 802.1Q header`](https://www.openvswitch.org/support/dist-docs-2.5/ovs-ofctl.8.txt), (vlan_tci is 16 bits in total 12 bits for vlan and 4 for priority + one extra bit). This also resembles the concept of a native vlan that network operators are familiar with as well. Maybe this is something that we could consider and see if it can simplify on our end.

3) `sdntrace_cp` relies on stored flows, so whatever is stored on `dl_vlan` is important to sdntrace_cp. @gretelliz will also need to adapt `sdntrace_cp`. Maybe we could always have a mask all 1s (0xfff) if it's not set? So this would also be aligned with vlan range? Something like str `"int/mask"`. We don't expect to have direct queries on the `flows.flow.match.dl_type` document attribute. So not storing the result int might not be an issue, and having it as str `"int/mask"`? would also facilitate for sdntrace_cp? Something to think about.

4) Get more insights about this functionality exploring on OpenFlow switches first. I'd also recommend to explore on OvS (and later on NoviFlow) to install `dl_vlan` flows that means `untagged` or `any` there and see how they ended up shown in the control-plane (`dump-flows` on OvS for instance) and generate some traffic. Having this data and information at least on OvS initially would be very helpful for us to ensure that our implementation end up installing equivalent flows. 

5) Double check and document which modes are supported to be installed in the same UNI, probably check with @italovalcy the valid combinations that we'd need to support in prod, for instance, should it be allowed to have an EPL and untagged in the same interface? Also check the default priorities flows implications we want to set. 

cc'ing also @ajoaoff since he'll work on mef eline vlan range

**Comment by Alopalao** [comment]
Commit [eac6bf0fdb30997a9d82e06a3cecabd2ba3b4699](https://github.com/kytos-ng/mef_eline/pull/258/commits/eac6bf0fdb30997a9d82e06a3cecabd2ba3b4699):
Added more possible strings to `value` from TagDoc, now it allows strings like `a/b` where:
- a and b are numbers, each have a maximum value of 4097

File: tests/unit/models/test_evc_deploy.py
Comment (Alopalao) [inline tests/unit/models/test_evc_deploy.py:321]: Hey @italovalcy I have one question here, what should happen to the flow action `set_vlan` when only one uni have the special case of `untagged` or `any`?. Right now `set_vlan` only is dismissed when the `in_vlan_a` is `0` or `"4096/4096"`. This last `elif` statemen does not delete/dismiss `{'action_type': 'set_vlan', 'vlan_id': out_vlan_a},`

File: tests/unit/models/test_evc_deploy.py
Comment (Alopalao) [inline tests/unit/models/test_evc_deploy.py:1363]: Same here @italovalcy. I added `set_vlan` here since the output should match 100 (not sure). Should it not set the vlan value to the packet here?

**Comment by Alopalao** [comment]
Commit [6f82d45628cd580878c039930d836fbdbc1fd357](https://github.com/kytos-ng/mef_eline/pull/258/commits/6f82d45628cd580878c039930d836fbdbc1fd357)
Closes #282 
The result of the EVC combinations can be seen in this [google sheet].(https://docs.google.com/spreadsheets/d/1QNZJrB163oqgsaZAtEebsywmRlMlzQzhRNBNtC68iWE/edit#gid=0)

File: tests/unit/models/test_evc_deploy.py
Comment (Alopalao) [reply tests/unit/models/test_evc_deploy.py]: Handled on commit [6f82d45628cd580878c039930d836fbdbc1fd357](https://github.com/kytos-ng/mef_eline/pull/258/commits/6f82d45628cd580878c039930d836fbdbc1fd357)

**Comment by viniarck** [comment]
@Alopalao the diffs of this PR is showing changes that have already been merged, I'll suggest you to try to `git rebase master`. Aldo, I'll wait for the UNI flow priority discussion of this comment to be confirmed/sorted out https://github.com/kytos-ng/mef_eline/issues/219#issuecomment-1452234821. Appreciated you managing to advance the code parts that were already clear.

File: tests/unit/models/test_evc_deploy.py
Comment (viniarck) [inline tests/unit/models/test_evc_deploy.py:1157]: - You can use a func to set the priority accordingly here too that was mentioned in the review. Could you use it here too to simplify?
- Could you also instead of `for` statements with `assertions` transform this into `pytest` `parametrize` or `with self.subTest`? parametrizing `uni_a` and `uni_z` accordingly? `parametrize` decorator is preferred, but since `mef_eline` still has tech debt using `unittest.TestCase` (and this is incompatible with `parametrize`) then it's OK to use `subTest` too whichever is easier.

**Review by viniarck — CHANGES_REQUESTED** [review]
This feature turned out really great @Alopalao, congrats.

I've asked small changes though. Also, could you update the changelog accordingly?

File: db/models.py
Comment (viniarck) [inline db/models.py:51]: Shouldn't we remove the `"n/n"` part from the message? and then leave for vlan range to take from this point forward?

File: controllers/__init__.py
Comment (Alopalao) [reply controllers/__init__.py:82]: Got it.

File: models/evc.py
Comment (Alopalao) [reply models/evc.py:1049]: Ok, I created a function that handles it.

File: openapi.yml
Comment (Alopalao) [reply openapi.yml:609]: I didn't know about that option. Added.

File: db/models.py
Comment (Alopalao) [reply db/models.py:51]: Yes, that is a leftover, removed.

File: models/evc.py
Comment (viniarck) [reply models/evc.py]: Nice. It looks really good. Let me close this thread.

File: openapi.yml
Comment (viniarck) [reply openapi.yml:609]: Right. It looks good, let me close this thread. 

cc'ing @ajoaoff for his info about vlan range to evolve from point forward in the discussion that's he's leading.

File: tests/unit/models/test_evc_deploy.py
Comment (viniarck) [reply tests/unit/models/test_evc_deploy.py]: Looks good. Let me close this.

File: db/models.py
Comment (viniarck) [reply db/models.py]: Got it. Let's close this thread then.

File: models/evc.py
Comment (viniarck) [inline models/evc.py:834]: Reading the PR again, I noticed that this tuple `{None, "4096/4096", 0}` appears six times (counting on tests too), it'll likely not change, so probably won't result in potential refactoring typo issues in the future, but if you manage to replace this as a constant with a name that's meaningful for the developer reading it that'd be even better.

**Comment by viniarck** [comment]
Aldo, the chagelog aldo needs to be updated if you could updated it. 

I've just reviewed your e2e tests too, nicely done. 

Once these minor suggestions here are addressed I'll approve this PR, and once @italovalcy  approves the e2e too since he was also interested in ensuring some flows being asserted then we'll merge this PR here OK?

File: models/evc.py
Comment (viniarck) [reply models/evc.py]: Looks good. Let me close this thread.

File: main.py
Comment (viniarck) [inline main.py:236]: @Alopalao could you also apply this same exc handling for the `PATCH /v2/evc/<circuit_id>` route? 

It turns out that when it's calling `evc.update` that will also call `evc.sync()` it's not handling `ValidationError` but only `ValueError` yet. 

We're still vulnerable to other validation issues there, sicne the `@validate` OpenAPI dec isn't applied there yet (this one here would be covered in a future oportunity)

File: models/evc.py
Comment (viniarck) [inline models/evc.py:1112]:
```suggestion
        if in_vlan not in self.special_cases and new_c_vlan == 0:
```

File: models/evc.py
Comment (viniarck) [inline models/evc.py:1107]:
```suggestion
        if new_c_vlan not in self.special_cases:
```

File: models/evc.py
Comment (viniarck) [inline models/evc.py:1124]:
```suggestion
        elif not in_vlan and new_c_vlan not in self.special_cases:
```

File: main.py
Comment (Alopalao) [reply main.py:236]: Got it.

File: models/evc.py
Comment (Alopalao) [reply models/evc.py]: I applied all of these suggestions in the last commit [5c98c93c79bffd79edb81fb824d46250e4db683e](https://github.com/kytos-ng/mef_eline/pull/258/commits/5c98c93c79bffd79edb81fb824d46250e4db683e)

File: models/evc.py
Comment (Alopalao) [reply models/evc.py]: Commit [5c98c93c79bffd79edb81fb824d46250e4db683e](https://github.com/kytos-ng/mef_eline/pull/258/commits/5c98c93c79bffd79edb81fb824d46250e4db683e)

Comment (viniarck) [reply models/evc.py]: Looks good, let me close the related threads.

**Review by viniarck — APPROVED** [review]
Nicely done, @Alopalao. Thanks for pushing these commits. I'll leave the PR approved, once e2e is also approved feel free to merge this PR. 

Also, when this is about to get merged, please notify on `#kytos-ng-dev` channel, that until [`sdntrace_cp` issue 79](https://github.com/kytos-ng/sdntrace_cp/issues/79) is addressed, this feature here is available, but after a controller restart consistency check will for any/untagged won't succeed and as a result, unnecessary periodic redeploys will end up happening, which is OK for now in `master` during development since that sdntrace_cp issue is also scheduled for `2023.1`, but let's make that clear for developers when they're using `master` version. Maybe we could have a marked `xfail` e2e test for this too, and once `sdntrace_cp` fixes it, then we remove the `@pytest.mark.xfail` decorator

**Pull request merged**
